tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "testlib/s2n_ktls_test_utils.h"
+
+/* These MOCK_IO* macros set errno before returning an error. These macros are
+ * mainly used for IO related operations (stuffer writes, NULL arguments,
+ * invalid length). Other sanity checks are guarded by the POSIX* macros.
+ *
+ * EINVAL will always be treated as a fatal errno code so its sufficient to
+ * use that as the generic errno code in the guard macros. */
+#define MOCK_IO_ENSURE(x)   \
+    do {                    \
+        if (!(x)) {         \
+            errno = EINVAL; \
+            return -1;      \
+        }                   \
+    } while (0)
+
+#define MOCK_IO_GUARD(x) \
+    MOCK_IO_ENSURE((x) > S2N_FAILURE)
+
+#define MOCK_IO_ENSURE_REF(x) \
+    MOCK_IO_ENSURE(S2N_OBJECT_PTR_IS_READABLE(x))
+
+/* Since its possible to read partial data, we need a way to update the length
+ * of the previous record for the mock stuffer IO implementation. */
+S2N_RESULT s2n_test_ktls_update_prev_header_len(struct s2n_test_ktls_io_stuffer *io_ctx, uint16_t remaining_len)
+{
+    RESULT_ENSURE_REF(io_ctx);
+    RESULT_ENSURE(remaining_len > 0, S2N_ERR_IO);
+
+    /* rewind the read ptr */
+    RESULT_GUARD_POSIX(s2n_stuffer_rewind_read(&io_ctx->ancillary_buffer, S2N_TEST_KTLS_MOCK_HEADER_LENGTH_SIZE));
+
+    /* rewrite the length */
+    uint8_t *ptr = s2n_stuffer_raw_read(&io_ctx->ancillary_buffer, S2N_TEST_KTLS_MOCK_HEADER_LENGTH_SIZE);
+    struct s2n_blob ancillary_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&ancillary_blob, ptr, S2N_TEST_KTLS_MOCK_HEADER_LENGTH_SIZE));
+    struct s2n_stuffer ancillary_stuffer = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&ancillary_stuffer, &ancillary_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_uint16(&ancillary_stuffer, remaining_len));
+
+    /* rewind to re-read the record with the remaining length */
+    RESULT_GUARD_POSIX(s2n_stuffer_rewind_read(&io_ctx->ancillary_buffer, S2N_TEST_KTLS_MOCK_HEADER_SIZE));
+
+    return S2N_RESULT_OK;
+}
+
+ssize_t s2n_test_ktls_sendmsg_stuffer_io(struct s2n_connection *conn, struct msghdr *msg, uint8_t record_type)
+{
+    POSIX_ENSURE_REF(conn);
```

Comment:
🤔 Shouldn't these calls also use MOCK_IO_REF? If the goal is to mock what sendmsg would return.

Reply:
I am going to get rid of these and use the POSIX macros instead. I was only using MOCK for IO related operations but it prob makes it harder to read.

Reply:
https://github.com/aws/s2n-tls/pull/4109/commits/f24d55ff425ed89f73d6ca13b33ea6511099064a

simplified since we only need to handle EAGAIN when reading/writing to the data_buffer